src/containers/color-indicator.jsx

```diff
@@ -0,0 +1,142 @@
+                isBitmap(this.props.format),
+                isStroke,
+                this.props.textEditTarget);
+            this._hasChanged = this._hasChanged || isDifferent;
+            this.props.onChangeColor(newColor, this.props.colorIndex);
+        }
+        handleChangeGradientType (gradientType) {
+            // Apply color and update redux, but do not update svg until picker closes.
+            const isDifferent = applyGradientTypeToSelection(
+                gradientType,
+                isBitmap(this.props.format),
+                isStroke,
+                this.props.textEditTarget);
+            this._hasChanged = this._hasChanged || isDifferent;
+            const hasSelectedItems = getSelectedLeafItems().length > 0;
+            if (hasSelectedItems) {
+                if (isDifferent) {
+                    // Recalculates the swatch colors
+                    this.props.setSelectedItems();
+                }
+            }
+            if (this.props.gradientType === GradientTypes.SOLID && gradientType !== GradientTypes.SOLID) {
+                // Generate color 2 and change to the 2nd swatch when switching from solid to gradient
+                if (!hasSelectedItems) {
+                    this.props.onChangeColor(getRotatedColor(this.props.color), 1);
+                }
+                this.props.onChangeColorIndex(1);
+            }
+            if (this.props.onChangeGradientType) this.props.onChangeGradientType(gradientType);
+        }
+        handleCloseColor () {
+            // If the eyedropper is currently being used, don't
+            // close the color menu.
+            if (this.props.isEyeDropping) return;
+
+            // Otherwise, close the color menu and
+            // also reset the color index to indicate
+            // that `color1` is selected.
+            this.props.onCloseColor();
+            this.props.onChangeColorIndex(0);
+        }
+        handleSwap () {
+            if (getSelectedLeafItems().length) {
+                const isDifferent = swapColorsInSelection(
+                    isBitmap(this.props.format),
+                    isStroke,
+                    this.props.textEditTarget);
+                this.props.setSelectedItems();
+                this._hasChanged = this._hasChanged || isDifferent;
+            } else {
+                let color1 = this.props.color;
+                let color2 = this.props.color2;
+                color1 = color1 === null || color1 === MIXED ? color1 : parseColor(color1).hex;
+                color2 = color2 === null || color2 === MIXED ? color2 : parseColor(color2).hex;
+                this.props.onChangeColor(color1, 1);
+                this.props.onChangeColor(color2, 0);
+            }
+        }
+        render () {
+            return (
+                <ColorIndicatorComponent
+                    {...this.props}
+                    label={this.props.intl.formatMessage(messages.label)}
```

Comment:
Could we pass in messages.label instead of messages?
Otherwise, I think we should document that label is expected to be defined on the messages passed in